main.py

```diff
@@ -31,11 +38,194 @@ def setup(self):
+            # Keys look like this: ('kytos', 'of_lldp')
+            if key[1] in self.subscribed_napps:
+                enable_napps.add(key[1])
+        return enable_napps
+
+    def start_enabling_pipeline(self, content: dict):
+        """Method to start the process to enable table
+        First, send event notifying NApps about their
+        new table set up.
+        """
+        name = "enable_table"
+        self.emit_event(name, content)
+
+    def build_content(self, pipeline: dict):
+        """Build content to be sent through an event"""
+        content = {}
+        with self._pipeline_lock:
+            for table in pipeline["multi_table"]:
+                table_id = table["table_id"]
+                for napp in table["napps_table_groups"]:
+                    if napp not in content:
+                        content[napp] = {}
+                    for flow_type in table["napps_table_groups"][napp]:
+                        content[napp][flow_type] = table_id
+        return content
+
+    def emit_event(self, name: str, content: dict = None):
+        """Send event"""
+        context = "kytos/of_multi_table"
+        event_name = f"{context}.{name}"
+        event = KytosEvent(name=event_name, content=content)
+        self.controller.buffers.app.put(event)
+
+    @listen_to("kytos/(mef_eline|coloring|of_lldp).enable_table")
+    def on_enable_table(self, event):
+        """Listen for NApps responses"""
+        self.handle_enable_table(event)
+
+    def handle_enable_table(self, event):
+        """Handle NApps responses from enable_table
+        Second, wait for all the napps to respond"""
+        napp = event.name.split('/')[1].split('.')[0]
+        # Check against the last current table
+        self.required_napps.remove(napp)
+        if self.required_napps:
+            # There are more required napps, 'waiting' responses
+            return
+        self.get_flows_to_be_installed()
+
+    def get_flows_to_be_installed(self):
+        """Get flows from flow manager so this NApp can modify them
+        Third, install the flows with different table_id"""
+        pipeline = self.get_enabled_table()
+        if pipeline.get("status") == "enabled":
+            # Pipeline enabling has finnished already
+            return
+        # Keys necessary to delete a flow, 'cookie_mask' is added later
+        keys = {'idle_timeout', 'hard_timeout', 'cookie', 'priority',
+                'match', 'actions'}
+        command = "v2/stored_flows?state=installed"
+        response = requests.get(f"{FLOW_MANAGER_URL}/{command}")
+
+        if response.status_code // 100 != 2:
+            log.error(f"Could not get the flows from flow_mager. Status "
+                      f"code {response.status_code}.")
+            # Problem here: pipeline stays as 'enabling', error needs
+            # to be handled. Events have been sent and NApps are probably
+            # working in a different pipeline setup
```

Comment:
Error here is not handled. I have two solutions for this:

- I can get the flows before sending any `enable_table` events and stored them in a self variable. Delete the content of the self variable after a pipeline has finished enabling or an error occurs.

- This part of the code is reached when enabling pipeline and disabling pipeline. 

1. Enabling pipeline error is easy to handle, resend `enable_table` with default pipeline and add extra value `error` to `self.required_napps`.

2. When disabling a pipeline, the NApp needs to return to default. I would need to get the most recent disabled pipeline enable it again and resend `enable_table` with extra value `error` to `self.required_napps`. 

The extra value is added so `get_flows_to_be_installed()` is not called, see lines 123-127.

Reply:
Very well thought @Alopalao. 

> Enabling pipeline error is easy to handle, resend enable_table with default pipeline

This sounds really great to me, it would make this NApp very reliable, since we know for a fact that falling back to the default pipeline would ensure a smooth operation. 

> When disabling a pipeline, the NApp needs to return to default. I would need to get the most recent disabled pipeline enable it again and resend enable_table with extra value error to self.required_napps.

Theoretically, if a pipeline was successfully enabled, when disabling it you cannot get any of the unrecoverable errors from opft_error assuming of course we'll only send coherent payload, so when disabling a pipeline always returning to the default sounds good to me. 

> The extra value is added so get_flows_to_be_installed() is not called, see lines 123-127.
I came with these solutions while writing this comment so maybe there is a better one.

Yes, there is @Alopalao . If instead of using requests you use `kytos.flow_manager.flows.(install|delete)` then it become simpler to avoid to deal with retries or timeouts, also you can leverage the `force=True` option, which will make `flow_manager` to deal with any socket error that can happen and eventually the consistency check will install or remove it. The only error you'd need to handle would be subscribing to `kytos/flow_manager.flow.error`, and since you'd be using `force=True`, then on `kytos/flow_manager.flow.error`, you only need to check if ```error_exception`` key is set, if it is then it's an OFPT_ERROR, so it's something you can't recover from and you'll need to rollback again to the basic pipeline. Also, another benefit of using events to install or delete flows is that it'd be easier to make it production-grade, notice that you can be dealing with 1000s of flows here which currently the responsibility is on the callers to pace it accordingly not to overwhelm the switches, with that said I'd recommend that you stick with a batched/interval approach [like mef_eline does grouping by switches](https://github.com/kytos-ng/mef_eline/blob/master/main.py#L729-L746), so to send for instance 1000 flows it would take 10 secs more or less, which is reasonable since installing a pipeline isn't keep happening frequently, and in fact we can expect to deal with 1000 ish flows in prod in this iteration, and mef_eline has also tested these parameters heavily on convergence tests so we know that switches can handle this sort of load without getting too overwhelemed, so could you also replace requests with the event interface when sending flows and parametrizing the batch_size/interval like mef_eline does, and later on we might further tune these parameters?

Reply:
Added mef_eline method to group flows by batches.
Also, added the handling of the defined error. It returns successfully to a previous state. Only one thing, if a user creates a napp called `error`, and a listener is added for `enable_table` this solution will not work and it will enter into a loop of trying to get flows from flow manager until a 200 is returned. Do we have a list of not allowed names for NApps?